### PR TITLE
Use CommandLineParser package for handling CLI args

### PR DESCRIPTION
**Pull request by ploughpuff**
I've been working on this PR to address my issue #693 

It uses package CommandLineParser:
https://github.com/commandlineparser/commandline

I tested out a few parsers and this one appears to be popular and still actively maintained.  It automatically provides a help and versions page, and moves the our CLI switches to be POSIX standard.

Please review and provide your thoughts.

Example output:
```
ploughpuff@vb-deb-jelly:~/jellyfin$ ~/tmp/bin/jellyfin --help
Jellyfin.Server 10.1.0.0
Copyright ©  2019 Jellyfin Contributors. Code released under the GNU General Public License Version 2

  -d, --programdata    Path to use for program data (databases files etc.).

  -c, --configdir      Path to use for config data (user policies and puctures).

  -l, --logdir         Path to use for writing log files.

  --ffmpeg             Path to external FFmpeg exe to use in place of built-in.

  --ffprobe            ffmpeg and ffprobe switches must be supplied together.

  --service            Run as headless service.

  --noautorunwebapp    Run headless if startup wizard is complete.

  --package-name       Used when packaging Jellyfin (example, synology).

  --restartpath        Path to reset script.

  --restartargs        Arguments for restart script.

  --help               Display this help screen.

  --version            Display version information.

```

### CODE REVIEW

**Comment by joshuaboniface** [comment]
I like it! It's gonna need some tight coordination with PRs to fix up the packaging defaults, but that shouldn't be too bad. My only concern is backwards compatibility - is there a way to include the single-dashed options too, at least for now, and strip them out later?

Also, if we're changing stuff:

>  -d, --programdata    Path to use for program data (databases files etc.).

Can we please rename this to "datadir" now? I hate that it's inconsistent with the others ;-)

**Comment by ploughpuff** [comment]
Second change-set added to provide backwards compatibility by regex replacing single-hyphens with double-hyphens, taking care not to replace single switches (like -d).

Renamed -programdatadir to -datadir.

File: Emby.Server.Implementations/Emby.Server.Implementations.csproj [hunk 1]
```diff
@@ -22,6 +22,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="CommandLineParser" Version="2.4.3" />
```
Comment (Bond-009) [inline Emby.Server.Implementations/Emby.Server.Implementations.csproj]: This should be moved to `Jellyfin.Server`

File: Emby.Server.Implementations/StartupOptions.cs [hunk 2]
```diff
@@ -1,30 +1,43 @@
-using System;
-using System.Linq;
-
 namespace Emby.Server.Implementations
 {
+    using CommandLine;
```
Comment (Bond-009) [inline Emby.Server.Implementations/StartupOptions.cs]: Using directives outside namespaces

File: Emby.Server.Implementations/StartupOptions.cs [hunk 3]
```diff
@@ -1,30 +1,43 @@
...
+
+    /// <summary>
+    /// Class used by CommandLine package when parsing the command line arguments.
+    /// </summary>
     public class StartupOptions
```
Comment (Bond-009) [inline Emby.Server.Implementations/StartupOptions.cs]: Move this whole class to `Jellyfin.Server`

File: Emby.Server.Implementations/StartupOptions.cs [hunk 4]
```diff
@@ -1,30 +1,43 @@
...
     {
-        private readonly string[] _options;
+        [Option('d', "datadir", Required = false, HelpText = "Path to use for the data folder (databases files etc.).")]
+        public string PathData { get; set; }
```
Comment (Bond-009) [inline Emby.Server.Implementations/StartupOptions.cs]: Ok, so this one is confusing since there is a dir `data` inside this one.

File: Emby.Server.Implementations/StartupOptions.cs [hunk 5]
```diff
@@ -1,30 +1,43 @@
...
+
+        [Option('c', "configdir", Required = false, HelpText = "Path to use for config data (user policies and puctures).")]
+        public string PathConfig { get; set; }
```
Comment (Bond-009) [inline Emby.Server.Implementations/StartupOptions.cs]: `ConfigDir` or `ConfigPath`

File: Emby.Server.Implementations/StartupOptions.cs [hunk 6]
```diff
@@ -1,30 +1,43 @@
...
+
+        [Option('l', "logdir", Required = false, HelpText = "Path to use for writing log files.")]
+        public string PathLog { get; set; }
+
+
+        [Option("ffmpeg", Required = false, HelpText = "Path to external FFmpeg exe to use in place of built-in.")]
+        public string FFmpeg { get; set; }
```
Comment (Bond-009) [inline Emby.Server.Implementations/StartupOptions.cs]: `FFmpegPath`

File: Emby.Server.Implementations/StartupOptions.cs [hunk 7]
```diff
@@ -1,30 +1,43 @@
...
+
+        [Option("ffprobe", Required = false, HelpText = "ffmpeg and ffprobe switches must be supplied together.")]
+        public string FFprobe { get; set; }
```
Comment (Bond-009) [inline Emby.Server.Implementations/StartupOptions.cs]: same

File: Emby.Server.Implementations/StartupOptions.cs [hunk 8]
```diff
@@ -1,30 +1,43 @@
...
+
+
+        [Option("service", Required = false, HelpText = "Run as headless service.")]
+        public bool Service { get; set; }
```
Comment (Bond-009) [inline Emby.Server.Implementations/StartupOptions.cs]: `IsSerivce`

File: Emby.Server.Implementations/StartupOptions.cs [hunk 9]
```diff
@@ -1,30 +1,43 @@
...
 
-        public StartupOptions(string[] commandLineArgs)
-        {
-            _options = commandLineArgs;
-        }
+        [Option("noautorunwebapp", Required = false, HelpText = "Run headless if startup wizard is complete.")]
+        public bool NoAutoRunWebApp { get; set; }
```
Comment (Bond-009) [inline Emby.Server.Implementations/StartupOptions.cs]: `AutoRunWebApp` although this might break backwards compat.

File: Jellyfin.Server/Program.cs [hunk 10]
```diff
@@ -26,6 +26,9 @@
 
 namespace Jellyfin.Server
 {
+    using CommandLine;
```
Comment (Bond-009) [inline Jellyfin.Server/Program.cs]: Using directives outside namespaces.

File: Jellyfin.Server/Program.cs [hunk 11]
```diff
@@ -35,14 +38,30 @@ public static class Program
...
         public static async Task Main(string[] args)
         {
-            StartupOptions options = new StartupOptions(args);
-            Version version = Assembly.GetEntryAssembly().GetName().Version;
-
-            if (options.ContainsOption("-v") || options.ContainsOption("--version"))
+            // For backwards compatibility.
+            // Modify any input arguments now which start with single-hyphen to POSIX standard
+            // double-hyphen to allow parsing by CommandLineParser package.
+            var pattern = @"^(-[^-\s]{2})"; // Match -xx, not -x, not --xx, not xx
```
Comment (Bond-009) [inline Jellyfin.Server/Program.cs]: Inline it or make it a const.

File: Emby.Server.Implementations/Emby.Server.Implementations.csproj (hunk 1 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/Emby.Server.Implementations.csproj]: Ok

File: Emby.Server.Implementations/StartupOptions.cs (hunk 4 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/StartupOptions.cs]: Yes bit confusing having that 'data' folder under -programdatadir.  What to do?  Rename this to DataDir?

File: Emby.Server.Implementations/StartupOptions.cs (hunk 5 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/StartupOptions.cs]: ConfigDir.  Change log to LogDir.  Then they all better match switch names.

File: Emby.Server.Implementations/StartupOptions.cs (hunk 7 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/StartupOptions.cs]: same?  I though the two FF switches where ok..  Change how please?

File: Emby.Server.Implementations/StartupOptions.cs (hunk 8 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/StartupOptions.cs]: Ok

File: Emby.Server.Implementations/StartupOptions.cs (hunk 9 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/StartupOptions.cs]: I can change the set to invert the switch, so we accept --noauthrunwebapp, but store AutoRunWebApp.  Is that what you want?

File: Jellyfin.Server/Program.cs (hunk 10 above)
Comment (ploughpuff) [reply Jellyfin.Server/Program.cs]: Ok will move outside namespace.  I blame VS17 ;)

File: Jellyfin.Server/Program.cs (hunk 11 above)
Comment (ploughpuff) [reply Jellyfin.Server/Program.cs]: Ok

File: Emby.Server.Implementations/StartupOptions.cs (hunk 7 above)
Comment (Bond-009) [reply Emby.Server.Implementations/StartupOptions.cs]: See the comment above.
`FFprobePath`

File: Emby.Server.Implementations/StartupOptions.cs (hunk 9 above)
Comment (Bond-009) [reply Emby.Server.Implementations/StartupOptions.cs]: Yep

File: Emby.Server.Implementations/StartupOptions.cs (hunk 7 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/StartupOptions.cs]: Ok got you

**Comment by ploughpuff** [comment]
Third change-set added to address Bond's review comments.

File: Emby.Server.Implementations/IStartupOptions.cs [hunk 12]
```diff
@@ -0,0 +1,55 @@
+namespace Emby.Server.Implementations.ParsedStartupOptions
```
Comment (Bond-009) [inline Emby.Server.Implementations/IStartupOptions.cs]: Make the namespace the same as the file path ie: `Emby.Server.Implementations`

File: Jellyfin.Server/StartupOptions.cs [hunk 13]
```diff
@@ -1,43 +1,47 @@
...
-
+        public string FFprobePath { get; set; }
 
         [Option("service", Required = false, HelpText = "Run as headless service.")]
-        public bool Service { get; set; }
+        public bool IsService { get; set; }
 
         [Option("noautorunwebapp", Required = false, HelpText = "Run headless if startup wizard is complete.")]
-        public bool NoAutoRunWebApp { get; set; }
+        public bool AutoRunWebApp { get => !NoautoRunWebApp; set => NoautoRunWebApp = value; }
```
Comment (Bond-009) [inline Jellyfin.Server/StartupOptions.cs]: This is a dirty hack.

File: Emby.Server.Implementations/IStartupOptions.cs [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+{
+    public interface IStartupOptions
+    {
+        /// <summary>
+        /// --datadir
+        /// </summary>
+        string DataDir { get; }
```
Comment (Bond-009) [inline Emby.Server.Implementations/IStartupOptions.cs]: The `Dir` ones don't need to be in the interface IMO, you can't be sure other server implementations will use them.

File: Jellyfin.Server/StartupOptions.cs (hunk 13 above)
Comment (ploughpuff) [reply Jellyfin.Server/StartupOptions.cs]: I played with alternatives but VS preferred this one.  What do you suggest?

File: Emby.Server.Implementations/IStartupOptions.cs (hunk 14 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/IStartupOptions.cs]: Ok, I'll remove from interface file.

File: Emby.Server.Implementations/IStartupOptions.cs (hunk 12 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/IStartupOptions.cs]: Ok

**Comment by ploughpuff** [comment]
Forth change-set added to address Bond's second set of review comments.

File: Emby.Server.Implementations/StartupOptions.cs (hunk 9 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/StartupOptions.cs]: On reflection, we decided not to make this change and stay with the NoAutoRunWebApp name.

File: Jellyfin.Server/StartupOptions.cs (hunk 13 above)
Comment (ploughpuff) [reply Jellyfin.Server/StartupOptions.cs]: On reflection, we decided to revert this change and stay with NoAutoRunWebApp.

File: Emby.Server.Implementations/StartupOptions.cs (hunk 4 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/StartupOptions.cs]: I think the solution is to rename the 'data' folder under DataDir to something else which better reflects its contents.  'database' might be better but that doesn't cover all contents.  May I suggest that improvement is not handled under this PR please?

File: Emby.Server.Implementations/StartupOptions.cs (hunk 6 above)
Comment (ploughpuff) [reply Emby.Server.Implementations/StartupOptions.cs]: Done.

File: Jellyfin.Server/Program.cs [hunk 15]
```diff
@@ -35,14 +37,30 @@ public static class Program
...
+            var regex = new Regex(pattern);
+
+            for (var i = 0; i < args.Length; i++)
             {
-                Console.WriteLine(version.ToString());
+                args[i] = regex.Replace(args[i], substitution);
             }
 
+            // For CommandLine package, change default behaviour to output errors to stdout (instead of stderr)
+            var parser = new Parser(config => config.HelpWriter = Console.Out);
```
Comment (JustAMan) [inline Jellyfin.Server/Program.cs]: hmm, why print errors to stdout?..

File: Jellyfin.Server/Program.cs [hunk 16]
```diff
@@ -78,7 +96,7 @@ public static async Task Main(string[] args)
                 Shutdown();
             };
 
-            _logger.LogInformation("Jellyfin version: {Version}", version);
+            _logger.LogInformation("Jellyfin version: {Version}", Assembly.GetEntryAssembly().GetName().Version);
```
Comment (JustAMan) [inline Jellyfin.Server/Program.cs]: I'm slightly wary of this change, plus it's not indicated in the PR. Please explain the reasoning

File: Jellyfin.Server/StartupOptions.cs [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+using Emby.Server.Implementations;
+
+namespace Jellyfin.Server
+{
+    /// <summary>
+    /// Class used by CommandLine package when parsing the command line arguments.
+    /// </summary>
+    public class StartupOptions : IStartupOptions
+    {
+        [Option('d', "datadir", Required = false, HelpText = "Path to use for the data folder (databases files etc.).")]
```
Comment (JustAMan) [inline Jellyfin.Server/StartupOptions.cs]:
```suggestion
        [Option('d', "datadir", Required = false, HelpText = "Path to use for the data folder (database files, etc.).")]
```

File: Jellyfin.Server/StartupOptions.cs [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+        public string DataDir { get; set; }
+
+        [Option('c', "configdir", Required = false, HelpText = "Path to use for config data (user policies and puctures).")]
```
Comment (JustAMan) [inline Jellyfin.Server/StartupOptions.cs]:
```suggestion
        [Option('c', "configdir", Required = false, HelpText = "Path to use for configuration data (user policies and pictures).")]
```
Btw, what's a "user policy"?

File: Jellyfin.Server/StartupOptions.cs [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+        public string ConfigDir { get; set; }
+
+        [Option('l', "logdir", Required = false, HelpText = "Path to use for writing log files.")]
+        public string LogDir { get; set; }
+
+        [Option("ffmpeg", Required = false, HelpText = "Path to external FFmpeg exe to use in place of built-in.")]
```
Comment (JustAMan) [inline Jellyfin.Server/StartupOptions.cs]:
```suggestion
        [Option("ffmpeg", Required = false, HelpText = "Path to external FFmpeg executable to use in place of default found in PATH.")]
```
or something similar... we don't have a built-in `ffmpeg`.

File: Jellyfin.Server/StartupOptions.cs [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+        public string FFmpegPath { get; set; }
+
+        [Option("ffprobe", Required = false, HelpText = "ffmpeg and ffprobe switches must be supplied together.")]
```
Comment (JustAMan) [inline Jellyfin.Server/StartupOptions.cs]: Help is not that much helpful... :) Maybe reword it somehow?

File: Jellyfin.Server/StartupOptions.cs [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+        [Option("service", Required = false, HelpText = "Run as headless service.")]
+        public bool IsService { get; set; }
+
+        [Option("noautorunwebapp", Required = false, HelpText = "Run headless if startup wizard is complete.")]
+        public bool NoAutoRunWebApp { get; set; }
+
+        [Option("package-name", Required = false, HelpText = "Used when packaging Jellyfin (example, synology).")]
+        public string PackageName { get; set; }
+
+        [Option("restartpath", Required = false, HelpText = "Path to reset script.")]
```
Comment (JustAMan) [inline Jellyfin.Server/StartupOptions.cs]:
```suggestion
        [Option("restartpath", Required = false, HelpText = "Path to restart script.")]
```

File: Jellyfin.Server/Program.cs (hunk 15 above)
Comment (ploughpuff) [reply Jellyfin.Server/Program.cs]: I thought stdout was sensible for some reason.  Remove and pipe to stderr (which is default)?

File: Jellyfin.Server/Program.cs (hunk 16 above)
Comment (ploughpuff) [reply Jellyfin.Server/Program.cs]: It provides the same output to log, just avoids the var creations.

File: Jellyfin.Server/Program.cs (hunk 15 above)
Comment (JustAMan) [reply Jellyfin.Server/Program.cs]: My thoughts are that errors should go to `stderr` if there's no strong reason for doing otherwise. End user can always redirect `stderr` to `stdout` if they desire so...

File: Jellyfin.Server/StartupOptions.cs (hunk 17 above)
Comment (ploughpuff) [reply Jellyfin.Server/StartupOptions.cs]: Ok

File: Jellyfin.Server/Program.cs (hunk 16 above)
Comment (JustAMan) [reply Jellyfin.Server/Program.cs]: I don't see you removing `version` variable anywhere, thus the question.

File: Jellyfin.Server/StartupOptions.cs (hunk 18 above)
Comment (ploughpuff) [reply Jellyfin.Server/StartupOptions.cs]: By user policy I mean policy.xml.  The xml which records your settings like EnableRemoteAccess, EnableContentDownloading, IsAdministrator etc.  

These help text strings are only my stab at making sense of what they do.  Please suggest better strings :)

File: Jellyfin.Server/StartupOptions.cs (hunk 19 above)
Comment (ploughpuff) [reply Jellyfin.Server/StartupOptions.cs]: Ok

File: Jellyfin.Server/StartupOptions.cs (hunk 20 above)
Comment (ploughpuff) [reply Jellyfin.Server/StartupOptions.cs]: My comment here was driven by observation in code where these switches are used.  If both --ffmpeg and --ffprobe are not provided, together, then which ever is given fails (silently) to be used.  Maybe we ignore that issue for now and revisit another time?

"Path to external FFprobe executable to use in place of default found in PATH."

File: Jellyfin.Server/StartupOptions.cs (hunk 21 above)
Comment (ploughpuff) [reply Jellyfin.Server/StartupOptions.cs]: Ok

File: Jellyfin.Server/StartupOptions.cs (hunk 20 above)
Comment (JustAMan) [reply Jellyfin.Server/StartupOptions.cs]: Maybe like this?
`Path to external FFprobe executable to use in place of default found in PATH. Must be specified along with FFmpeg`

File: Jellyfin.Server/Program.cs (hunk 15 above)
Comment (ploughpuff) [reply Jellyfin.Server/Program.cs]: Removed this line and lets use default behaviour.

File: Jellyfin.Server/StartupOptions.cs (hunk 18 above)
Comment (ploughpuff) [reply Jellyfin.Server/StartupOptions.cs]: Changed to user settings.

File: Jellyfin.Server/StartupOptions.cs (hunk 20 above)
Comment (ploughpuff) [reply Jellyfin.Server/StartupOptions.cs]: Changed to above.

**Comment by ploughpuff** [comment]
Fifth change-set added to address review comment by LikeAMan.

**Comment by joshuaboniface** [comment]
@Bond-009 It's really application-dependent, though maybe `-p` would be better. `-p/--datadir`

**Comment by JustAMan** [comment]
> @Bond-009 It's really application-dependent, though maybe `-p` would be better. `-p/--datadir`

I don't see any `p` in `datadir`, sorry :) seems strange to use it. I like `-d` more.

**Comment by joshuaboniface** [comment]
> > @Bond-009 It's really application-dependent, though maybe `-p` would be better. `-p/--datadir`
> 
> I don't see any `p` in `datadir`, sorry :) seems strange to use it. I like `-d` more.

Well they don't always have to match, but I'm really on the fence. I don't mind `-d`, especially since we don't really have a "fork and daemonize" in the codebase (and I don't think adding one would be worthwhile), so that's fine.
